doc/hcitop.1.rst
.. SPDX-FileCopyrightText: 2021-2025 BlueALSA developers
.. SPDX-License-Identifier: MIT

======
hcitop
======

-------------------------------------
a simple dynamic view of HCI activity
-------------------------------------

:Date: July 2023
:Manual section: 1
:Manual group: General Commands Manual
:Version: $VERSION$

SYNOPSIS
========

**hcitop** [*OPTION*]...

DESCRIPTION
===========

**hcitop** provides a dynamic real-time view of activity statistics for each
HCI interface. The view is refreshed at regular intervals, and also on demand
by pressing a key. To quit the program press the 'q' key, or use Ctrl-C.

OPTIONS
=======

-h, --help
    Output a usage message and exit.

-V, --version
    Output the version number and exit.

-d SEC, --delay=SEC
    Set the interval at which the statistics are refreshed. SEC is a number of
    seconds and may include a decimal point or exponent.

COLUMNS
=======

HCI
    The HCI name ("hci0", etc.).

BUS
    The bus name ("UART", "USB", etc.).

ADDR
    The Bluetooth device address.

FLAGS
    Status flags of the HCI. See ``FLAGS`` below.

RX
    Total amount of data received since the HCI was last brought up.

TX
    Total amount of data transmitted since the HCI was last brought up.

RX/s
    Average rate of reception during the last refresh interval.

TX/s
    Average rate of transmission during the last refresh interval.

FLAGS
=====

An array of flag characters indicating the current status of the HCI. The flags
are shown in the following order. The indicated letter appears when that flag
is "TRUE", a blank is shown when the flag is "FALSE".

U
    The interface is "Up".

N
    The interface is initializing.

R
    The interface is running.

P
    Page scan is enabled.

I
    Inquiry scan is enabled.

A
    Authentication is enabled.

E
    Encryption is enabled.

Q
    The interface is currently inquiring (scanning) remote devices.

X
    Raw mode is enabled.

SEE ALSO
========

``btmon(1)``, ``hciconfig(1)``, ``hcitool(1)``

Project web site
  https://github.com/arkq/bluez-alsa


NEWS
unreleased
==========

- renamed bluealsa to bluealsad (no backward compatibility)
- renamed bluealsa.conf to org.bluealsa.conf (no backward compatibility)
- renamed bluealsa-cli to bluealsactl (no backward compatibility)
- optional support for A2DP Sink and Source with LHDC v3 codec
- channel map and volume control for surround sound (5.1, 7.1) audio
- native A2DP volume control by default (dropped --a2dp-volume option)
- optional support for adaptive audio resampling in bluealsa-aplay
- fix configuration for Android 13 A2DP Opus codec
- improved ALSA PCM support for A2DP-sink, HFP-HF and HSP-HS
- support multiple client connections to each Bluetooth PCM

bluez-alsa v4.3.1 (2024-08-30)
==============================

- fix crash when playing audio with more than 2 channels
- fix AAC configuration selection for low sample rates

bluez-alsa v4.3.0 (2024-08-13)
==============================

- optional support for Android 13 A2DP Opus codec
- multi channels and sample rates mode for ALSA PCM plug-in
- bluealsa-aplay: fix volume synchronization on Raspberry Pi

bluez-alsa v4.2.0 (2024-05-11)
==============================

- optional support for BLE MIDI 1.0 profile as a GATT server
- optional support for HFP LC3-SWB (Super Wide Band) codec
- command line option to set real-time priority for IO threads
- command line option to filter log messages based on severity
- allow to select individual extended controls in ALSA plug-in
- codec-specific delay adjustment with ALSA control and persistency
- improved AAC codec support to be conformant with A2DP v1.4
- fix for SBC codec and audio scaling on big-endian platforms
- fix mSBC decode for MTU > 60 bytes (e.g. Realtek USB adapters)
- bluealsa-aplay: option to auto-select volume mode for used PCM
- support use of libfreeaptx for apt-X and apt-X HD codecs
- stability fixes for ALSA PCM I/O plug-in

bluez-alsa v4.1.1 (2023-06-24)
==============================

- fix for oFono HF role SCO socket authorization
- fix for mSBC MTU adjustment for Realtek USB adapters
- bluealsa-aplay: fix playback after BT codec update

bluez-alsa v4.1.0 (2023-05-23)
==============================

- removed deprecated org.bluealsa.Manager1 D-Bus interface
- persistent storage for PCM volume and mute state
- PCM volume control with oFono HFP-AG and HFP-HF profiles
- transport running state exported in PCM D-Bus interface
- A2DP codec configuration blob exported in PCM D-Bus interface
- optional non-dynamic operation mode for ALSA control plug-in
- optional extended controls for ALSA control plug-in
- changed RFCOMM D-Bus API features property to array of strings
- fix for SCO link establishment for oFono HFP-AG profile
- fix for volume control for HSP-HS and HFP-HF profiles
- stability fixes for ALSA PCM I/O and control plug-ins
- bluealsa-aplay: fix for volume synchronization
- lots of fixes for race conditions (TSAN)
- lots of updates to the manual pages

bluez-alsa v4.0.0 (2022-06-03)
==============================

- changes in command line options (no backward compatibility)
- implement D-Bus ObjectManager (deprecate org.bluealsa.Manager1)
- optional support for A2DP FastStream codec (music & voice)
- optional support for A2DP LC3plus codec (music & voice)
- enhanced SBC XQ mode (XQ+) which bumps bitrate to 595 kbps
- packet loss concealment (PLC) for HFP with mSBC codec
- enable/disable BT codecs via command line options
- allow to select BT transport codec with ALSA configuration
- allow to set PCM volume properties with ALSA configuration
- optional single-device mode for ALSA control plug-in
- export battery status/charge via BlueZ battery provider
- connection keep-alive timeout for all BT profiles
- better conformance with RTP specification for A2DP codecs
- fix for rapid consecutive SCO link close(2)/connect(2)
- bluealsa-aplay: forward PCM volume to ALSA mixer element
- systemd service files for bluealsa and bluealsa-aplay
- lots of fixes for ALSA PCM I/O and control plug-ins

bluez-alsa v3.1.0 (2021-06-01)
==============================

- higher PCM bit depth for apt-X HD (24-bit) and LDAC (32-bit)
- support for A2DP Sink with apt-X (HD) if decoder is available
- support for A2DP Sink with LDAC codec if decoder is available
- better audio volume scaling (Bluetooth volume as loudness)
- support TLV data for dB calculations in ALSA control plug-in
- fix PCM delay reporting from connected Bluetooth devices
- most recently connected Bluetooth device as a default PCM
- lots of improvements to the ALSA PCM I/O plug-in
- use rst2man (docutils) instead of pandoc to build man-pages
- bluealsa-cli: utility for using D-Bus API from command line
- bash completion script for installed applications

bluez-alsa v3.0.0 (2020-09-15)
==============================

- separate D-Bus paths for PCM sink and source (API breakage)
- store PCM sample physical width in the D-Bus format property
- list available and supported audio codecs via D-Bus API
- support for A2DP audio codec selection or reconfiguration
- fix SBC encoder bit-pool selection (support SBC XQ quality)
- allow to disable software volume scaling via D-Bus API
- great improvement to the PCM I/O plug-in (thanks to @borine)
- install bluez-alsa manual pages (optional, requires pandoc)
- bluealsa-aplay: list connected BT devices and available PCMs
- bluealsa-aplay: fix PCM corruption due to out-of-bounds read

bluez-alsa v2.1.0 (2020-01-25)
==============================

- support HSP/HFP headset mode (connect phone without oFono)
- optional integration with UPower for battery level reporting
- initial (HFP only) support for audio codec selection
- auto-detect BlueZ D-Bus service appearance/disappearance
- lots of improvements for SCO connection robustness
- route SCO data via transport interface in Broadcom chips
- lots of fixes for ALSA plug-ins and BlueALSA server

bluez-alsa v2.0.0 (2019-10-15)
==============================

- BlueALSA server as a D-Bus service with public API

utils/hcitop.c
/*
 * BlueALSA - hcitop.c
 * SPDX-FileCopyrightText: 2016-2025 BlueALSA developers
 * SPDX-License-Identifier: MIT
 */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <curses.h>
#include <bsd/stdlib.h>

#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>

static const struct {
	unsigned int bit;
	char flag;
} hci_flags_map[] = {
	{ HCI_UP, 'U' },
	{ HCI_INIT, 'N' },
	{ HCI_RUNNING, 'R' },
	{ HCI_PSCAN, 'P' },
	{ HCI_ISCAN, 'I' },
	{ HCI_AUTH, 'A' },
	{ HCI_ENCRYPT, 'E' },
	{ HCI_INQUIRY, 'Q' },
	{ HCI_RAW, 'X' },
};

static size_t get_devinfo(struct hci_dev_info di[HCI_MAX_DEV]) {
	size_t num = 0;
	for (size_t i = 0; i < HCI_MAX_DEV; i++)
		if (hci_devinfo(i, &di[num]) == 0)
			num++;
	return num;
}

static unsigned int get_average_rate(unsigned int *array, size_t size) {

	/* at least two points are required */
	if (size < 2)
		return 0;

	unsigned int x = 0;
	unsigned int y = 0;
	size_t i;

	size--;
	i = size;

	while (i--) {
		int b = (array[i] - array[i + 1]) % size;
		x += (array[i] - array[i + 1]) / size;
		if (y >= size - b) {
			y -= size - b;
			x++;
		}
		else {
			y += b;
		}
	}

	return x + y / size;
}

static void sprint_hci_flags(char *str, unsigned int flags) {

	size_t i;

	for (i = 0; i < sizeof(hci_flags_map) / sizeof(*hci_flags_map); i++)
		str[i] = hci_test_bit(hci_flags_map[i].bit, &flags) ? hci_flags_map[i].flag : ' ';

	str[i] = '\0';
}

int main(int argc, char *argv[]) {

	int opt;
	const char *opts = "hVd:";
	const struct option longopts[] = {
		{ "help", no_argument, NULL, 'h' },
		{ "version", no_argument, NULL, 'V' },
		{ "delay", required_argument, NULL, 'd' },
		{ 0, 0, 0, 0 },
	};

	int delay_sec = 1;
	int delay_msec = 0;

	while ((opt = getopt_long(argc, argv, opts, longopts, NULL)) != -1)
		switch (opt) {
		case 'h' /* --help */ :
			printf("usage: %s [ -d sec ]\n"
					"  -h, --help\t\tprint this help and exit\n"
					"  -V, --version\t\tprint version and exit\n"
					"  -d, --delay=SEC\tdelay time interval\n",
					argv[0]);
			return EXIT_SUCCESS;

		case 'V' /* --version */ :
			printf("%s\n", PACKAGE_VERSION);
			return EXIT_SUCCESS;

		case 'd' /* --delay=SEC */ :
			delay_sec = atoi(optarg);
			delay_msec = (int)((atof(optarg) - delay_sec) * 10) * 100;
			if (delay_sec < 0 || delay_msec < 0 || (delay_sec == 0 && delay_msec == 0)) {
				fprintf(stderr, "%s: -d requires positive argument (max precision: 0.1)\n", argv[0]);
				return EXIT_FAILURE;
			}
			break;

		default:
			fprintf(stderr, "Try '%s --help' for more information.\n", argv[0]);
			return EXIT_FAILURE;
		}

	struct hci_dev_info devices[HCI_MAX_DEV];
	unsigned int byte_rx[HCI_MAX_DEV][3];
	unsigned int byte_tx[HCI_MAX_DEV][3];

	memset(byte_rx, 0, sizeof(byte_rx));
	memset(byte_tx, 0, sizeof(byte_tx));

	initscr();
	cbreak();
	noecho();
	curs_set(0);

	for (size_t ii = 1;; ii++) {

		const char *template_top = "%-5s %4s %-17s %-9s %8s %8s %8s %8s";
		const char *template_row = "%-5s %4s %17s %9s %8s %8s %8s %8s";

		attron(A_REVERSE);
		mvprintw(0, 0, template_top, "HCI", "BUS", "ADDR", "FLAGS", "RX", "TX", "RX/s", "TX/s");
		attroff(A_REVERSE);

		size_t count = get_devinfo(devices);
		for (size_t i = 0; i < HCI_MAX_DEV; i++) {

			/* shift historic data to the right by one sample */
			memmove(&byte_rx[i][1], &byte_rx[i][0], sizeof(*byte_rx) - sizeof(**byte_rx));
			memmove(&byte_tx[i][1], &byte_tx[i][0], sizeof(*byte_tx) - sizeof(**byte_tx));

			if (i >= count)
				continue;

			char addr[18];
			ba2str(&devices[i].bdaddr, addr);

			const char *bus = hci_dtypetostr(devices[i].type);

			char flags[sizeof(hci_flags_map) / sizeof(*hci_flags_map) + 1];
			sprint_hci_flags(flags, devices[i].flags);

			byte_rx[i][0] = devices[i].stat.byte_rx;
			byte_tx[i][0] = devices[i].stat.byte_tx;

			const size_t ii_max = sizeof(*byte_rx) / sizeof(**byte_rx);
			const size_t samples = ii < ii_max ? ii : ii_max;
			unsigned int rate_rx = get_average_rate(byte_rx[i], samples);
			unsigned int rate_tx = get_average_rate(byte_tx[i], samples);

			rate_rx = rate_rx * 10 / (delay_sec * 10 + delay_msec / 100);
			rate_tx = rate_tx * 10 / (delay_sec * 10 + delay_msec / 100);

			char rx[7], rx_rate[9];
			char tx[7], tx_rate[9];

			humanize_number(rx, sizeof(rx), byte_rx[i][0], "B", HN_AUTOSCALE, 0);
			humanize_number(tx, sizeof(tx), byte_tx[i][0], "B", HN_AUTOSCALE, 0);
			humanize_number(rx_rate, sizeof(rx_rate), rate_rx, "B", HN_AUTOSCALE, 0);
			humanize_number(tx_rate, sizeof(tx_rate), rate_tx, "B", HN_AUTOSCALE, 0);

			mvprintw(i + 1, 0, template_row,
					devices[i].name, bus, addr, flags, rx, tx, rx_rate, tx_rate);

		}

		timeout(delay_sec * 1000 + delay_msec);
		if (getch() == 'q')
			break;

	}

	endwin();
	return EXIT_SUCCESS;
}
